lib/smart_proxy_dynflow_core/log.rb

```diff
@@ -54,13 +54,28 @@ def roll_log
       end
     end
 
+    class ProxyStructuredFormater < ::Dynflow::LoggerAdapters::Formatters::Abstract
+      def call(_severity, _datetime, _prog_name, message)
+        if ::Exception === message
+          @base.exception("Error details", message)
+          "#{message.message} (#{message.class})"
+        else
+          message
+        end
+      end
+
+      def format(message)
+        call(nil, nil, nil, message)
+      end
+    end
+
     class ProxyAdapter < ::Dynflow::LoggerAdapters::Simple
-      def initialize(logger, level = Logger::DEBUG, formatters = [::Dynflow::LoggerAdapters::Formatters::Exception])
+      def initialize(logger, level = Logger::DEBUG, formatters = [])
         @logger           = logger
         @logger.level     = level
-        @logger.formatter = method(:formatter).to_proc
-        @action_logger    = apply_formatters ProgNameWrapper.new(@logger, ' action'), formatters
-        @dynflow_logger   = apply_formatters ProgNameWrapper.new(@logger, 'dynflow'), formatters
+        @logger.formatter = ProxyStructuredFormater.new(@logger)
+        @action_logger    = apply_formatters(ProgNameWrapper.new(@logger, ' action'), [ProxyStructuredFormater])
```

Comment:
When calling `action_logger.info(RuntimeError.new("just a test"))`, it fails with

```
E, [2018-10-23T13:50:43.980012 #21350] ERROR -- /parallel-executor-core/pool default/worker-4: undefined method `exception' for #<Dynflow::LoggerAdapters::Simple::ProgNameWrapper:0x0055ab02321088>
/home/inecas/Projects/ws/foreman-rex/smart_proxy_dynflow/lib/smart_proxy_dynflow_core/log.rb:60:in `call'
/home/inecas/Projects/ws/foreman-rex/smart_proxy_dynflow/lib/smart_proxy_dynflow_core/log.rb:68:in `format'
```
That's probably because the `@base` in this case would not be the `@logger` but the `ProgNameWrapper`. I would need to think of the best way to solve this issue: any suggestions welcome

Reply:
Ok, part of it should be solved by this PR in dynlfow https://github.com/Dynflow/dynflow/pull/300, however, we should still add  `@base.exception("Error details", message) if @base.respond_to?(:exception)` to https://github.com/theforeman/smart_proxy_dynflow/pull/56/files#diff-fe07044a5ec271675cdb2f5043c6ca77R60, as when running smart_proxy_dynflow_core, the logger might not have the `exception` method defined

Reply:
Ok that works for me. Ideally we would move from Ruby logging stack which is really very simplified and limiting to logging gem in Dynflow with its powerful things like structured logging. Then you could even drop these wrappers completely because you can put things into logging contexts and have them appear in logs.